src/solvers/postSolver.ml

```diff
@@ -211,13 +215,30 @@ struct
       if M.tracing then M.trace "postsolver" "one_constraint %a %a\n" S.Var.pretty_trace x S.Dom.pretty rhs;
       PS.one_constraint ~vh ~x ~rhs
     in
-    List.iter one_var vs;
-
-    PS.finalize ~vh ~reachable;
+    (Stats.time "potentially_incremental_reach" (List.iter one_var)) vs;
+
+    match dep with
+    | None -> PS.finalize ~vh ~reachable:reachable;
+    | Some dep -> (
+        (* Perform reachability on whole constraint system, but cheaply by using logged dependencies *)
+        (* This only works if the other reachability has been performed before, so dependencies created only during postsolve are recorded *)
+        let reachable' = VH.create (VH.length vh) in
+        let rec one_var' x =
+          if not (VH.mem reachable' x) then (
+            VH.replace reachable' x ();
+            match VH.find_option dep x with
+            | Some vs -> VS.iter one_var' vs
+            | None -> ()
+          )
+        in
+        (Stats.time "cheap_full_reach" (List.iter one_var')) vs;
+        PS.finalize ~vh ~reachable:reachable';
```

Comment:
I didn't yet dive into this PR, but here's an idea which might minimize the amount of changes (and possibly require _none_ to the postsolver itself): do the `reachable'` computation based on `dep` directly in TD3 and use that as `init_reachable`.

Currently there are two possibilities for `init_reachable` in TD3:
1. Empty, so full everything is re-evaluated (so no incrementality).
2. `superstable`, so there's maximal reuse (which is too much for the issue in question).

But I think any set between the two would also be a completely fine choice and the cheaply computed reachability (or I guess the complement of it) should be a subset of `superstable`.

Reply:
> do the `reachable'` computation based on `dep` directly in TD3 and use that as `init_reachable`.

Interesting idea. However, I am a bit skeptical about this, as the point precisely is that during the computation of `reachable'` one has to descend into parts of the constraint system that TD3 does not have to consider. I think for this to work, we would have to store a set of all transitive(!) dependencies of an unknown with it during the non-incremental initial run and marshal that with the rest, such a set might be huge for larger programs so I am a bit worried that maintaining it might outweigh the savings from not having to descend during the computation of `reachable'`

Reply:
This `reachable'` computation is only based on `vs` and `dep`, both of which are directly provided by TD3 here. Seems to me like it could directly be copied to TD3 to do the same thing without this `one_var'` having to consider anything else that TD3 uses (`superstable`, etc).

Reply:
A crucial detail is that it needs to happen only after the post-solver has run as otherwise any global the receives values only during post-solving would be pruned away.
The other thing is that `reachable'` should not be supplied as `init_reachable`, as it includes also all new things, and the normal post-solver still needs to iterate over that to generate warnings/accesses.

Reply:
> A crucial detail is that it needs to happen only after the post-solver has run as otherwise any global the receives values only during post-solving would be pruned away.

If a global receives a value during postsolving, then it goes into the postsolver's `reachable` and hence should not be pruned.
Looks like here the `dep`-based `reachable'` completely replaces the `reachable`, which the postsolver is computing based on actual evaluations of the right-hand sides, so accounts correctly for all the `Goblintutil.postsolving` differences.


> The other thing is that `reachable'` should not be supplied as `init_reachable`, as it includes also all new things, and the normal post-solver still needs to iterate over that to generate warnings/accesses.

Hence my comment above about the complement. It should be possible to filter `superstable` using this `reachable'` (keep only those in both, giving a subset of `superstable`) and use that as `init_reachable`. Then all the new things would automatically be excluded since they aren't superstable, but we prevent the unnecessary parts of superstable from being reused.